streams/src/main/java/org/apache/kafka/streams/processor/internals/StateManagerUtil.java

```diff
@@ -38,13 +41,39 @@
  */
 final class StateManagerUtil {
     static final String CHECKPOINT_FILE_NAME = ".checkpoint";
+    static final long OFFSET_DELTA_THRESHOLD_FOR_CHECKPOINT = 10_000L;
 
     private StateManagerUtil() {}
 
     static RecordConverter converterForStore(final StateStore store) {
         return isTimestamped(store) ? rawValueToTimestampedValue() : identity();
     }
 
+    static boolean checkpointNeeded(final boolean enforceCheckpoint,
+                                    final Map<TopicPartition, Long> oldOffsetSnapshot,
+                                    final Map<TopicPartition, Long> newOffsetSnapshot) {
+        // we should always have the old snapshot post completing the register state stores;
+        // if it is null it means the registration is not done and hence we should not overwrite the checkpoint
+        if (oldOffsetSnapshot == null)
+            return false;
+
+        // if the previous snapshot is empty while the current snapshot is not then we should always checkpoint;
+        // note if the task is stateless or stateful but no stores logged, the snapshot would also be empty
+        // and hence it's okay to not checkpoint
+        if (oldOffsetSnapshot.isEmpty() && !newOffsetSnapshot.isEmpty())
+            return true;
+
+        // we can checkpoint if the the difference between the current and the previous snapshot is large enough
+        long totalOffsetDelta = 0L;
+        for (final Map.Entry<TopicPartition, Long> entry : newOffsetSnapshot.entrySet()) {
+            totalOffsetDelta += Math.abs(oldOffsetSnapshot.getOrDefault(entry.getKey(), 0L) - entry.getValue());
+        }
+
+        // when enforcing checkpoint is required, we should overwrite the checkpoint if it is different from the old one;
+        // otherwise, we only overwrite the checkpoint if it is largely different from the old one
+        return enforceCheckpoint ? totalOffsetDelta > 0 : totalOffsetDelta > OFFSET_DELTA_THRESHOLD_FOR_CHECKPOINT;
```

Comment:
I'm a bit torn about this optimization to avoid double checkpointing, because on the other hand, even if we have not made any progress since loaded the checkpoint, we'd still need to make a checkpoint upon closing if we have never made one before -- and I use emptyMap as an indicator for that.

Given that upon suspending we are now less likely to checkpoint, the chance that we would double checkpointing (when transiting to suspended, and when transiting to closed) is smaller, and hence I'm thinking maybe I'd just remove this optimization to make the logic a bit simpler. LMK WDYT.

Reply:
Sorry I meant the "commit" during suspension: when we suspend a task we would commit the offsets and in `postCommit` we may write the checkpoint file, and then when we close the suspended task we may commit again (note the `prepareCommit` call should return empty offsets since there's nothing new to commit, but in `postCommit` we would checkpoint the file again, i.e. double writing checkpoint.

So what I did here is to not enforce checkpointing in `postCommit` during suspension, and only enforce checkpointing in `postCommit` during closing.

Reply:
I'm still not following -- I thought we no longer commit during close at all, we just commit when suspending? For active tasks we call pre/postCommit and suspend in `handleRevocation` and then just close them in `handleAssignment`. For standbys we do all those things in `handleAssignment`, but still only once

Reply:
Oh, I see, we now do the checkpoint in `handleAssignment` as well. Why? It seems like we're just bringing back the problem of potentially double-checkpointing that had been solved?

Reply:
Can you elaborate on why not enforce the checkpoint during suspension in `handleRevocation` (for the tasks to be closed)? It seems like we just re-introduce this problem of potentially double checkpointing and I'm not sure I understand why we need to checkpoint only in `handleAssignment`. Is this an optimization or a correctness issue or a code cleanness thing?

Reply:
I think we do not need to enforce checkpoint during suspension but only need to do that during closure / recycling; if a suspended task is resumed then we do not need to write checkpoint in between.

But admittedly moving forward most suspended tasks would be closed or recycled :slightly_smiling_face: So I can change that back.

Reply:
In the current code we actually call `postCommit(true)` both during suspension and closure. It's just that in `checkpointNeeded` we would only return true with `enforce` if it has changed --- personally I think this is a general optimization worth applying.